File: app/src/main/java/com/example/bloodbank/CompatibleUsersActivity.java
package com.example.bloodbank;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.bloodbank.Adapter.CompatibleUserAdapter;
import com.example.bloodbank.Model.CompatibleUser;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

public class CompatibleUsersActivity extends AppCompatActivity {
    private RecyclerView recyclerView;
    private CompatibleUserAdapter adapter;
    private List<CompatibleUser> userList;
    private ProgressBar progressBar;
    private TextView noUsersText;
    private DatabaseReference usersRef;
    private String currentUserBloodGroup;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_category_selected);

        // Initialize views
        recyclerView = findViewById(R.id.recyclerView);
        progressBar = findViewById(R.id.progressBar);
        noUsersText = findViewById(R.id.noUsersText);
        userList = new ArrayList<>();
        adapter = new CompatibleUserAdapter(this, userList);

        // Setup RecyclerView
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(adapter);

        // Get current user's blood group
        String currentUserId = FirebaseAuth.getInstance().getCurrentUser().getUid();
        usersRef = FirebaseDatabase.getInstance().getReference().child("users");

        // First, get current user's blood group
        usersRef.child(currentUserId).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                if (snapshot.exists()) {
                    currentUserBloodGroup = snapshot.child("bloodGroup").getValue(String.class);
                    if (currentUserBloodGroup != null) {
                        fetchCompatibleUsers();
                    }
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                Log.e("CompatibleUsers", "Error fetching current user data: " + error.getMessage());
                progressBar.setVisibility(View.GONE);
                noUsersText.setVisibility(View.VISIBLE);
            }
        });
    }

    private void fetchCompatibleUsers() {
        progressBar.setVisibility(View.VISIBLE);
        noUsersText.setVisibility(View.GONE);

        usersRef.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                userList.clear();
                String currentUserId = FirebaseAuth.getInstance().getCurrentUser().getUid();

                for (DataSnapshot userSnapshot : snapshot.getChildren()) {
                    String userId = userSnapshot.getKey();
                    if (userId != null && !userId.equals(currentUserId)) {
                        String bloodGroup = userSnapshot.child("bloodGroup").getValue(String.class);
                        String type = userSnapshot.child("type").getValue(String.class);
                        
                        if (bloodGroup != null && type != null && isCompatible(bloodGroup)) {
                            CompatibleUser user = new CompatibleUser(
                                userId,
                                userSnapshot.child("name").getValue(String.class),
                                type,
                                bloodGroup,
                                userSnapshot.child("profilePictureUrl").getValue(String.class),
                                userSnapshot.child("email").getValue(String.class),
                                userSnapshot.child("phoneNumber").getValue(String.class),
                                userSnapshot.child("address").getValue(String.class),
                                userSnapshot.child("lastDonation").getValue(String.class)
                            );
                            userList.add(user);
                        }
                    }
                }

                progressBar.setVisibility(View.GONE);
                if (userList.isEmpty()) {
                    noUsersText.setVisibility(View.VISIBLE);
                } else {
                    noUsersText.setVisibility(View.GONE);
                    adapter.updateList(userList);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                Log.e("CompatibleUsers", "Error fetching users: " + error.getMessage());
                progressBar.setVisibility(View.GONE);
                noUsersText.setVisibility(View.VISIBLE);
            }
        });
    }

    private boolean isCompatible(String donorBloodGroup) {
        // Define blood group compatibility rules
        switch (currentUserBloodGroup) {
            case "A+":
                return donorBloodGroup.equals("A+") || donorBloodGroup.equals("A-") || 
                       donorBloodGroup.equals("O+") || donorBloodGroup.equals("O-");
            case "A-":
                return donorBloodGroup.equals("A-") || donorBloodGroup.equals("O-");
            case "B+":
                return donorBloodGroup.equals("B+") || donorBloodGroup.equals("B-") || 
                       donorBloodGroup.equals("O+") || donorBloodGroup.equals("O-");
            case "B-":
                return donorBloodGroup.equals("B-") || donorBloodGroup.equals("O-");
            case "AB+":
                return true; // AB+ can receive from all blood groups
            case "AB-":
                return donorBloodGroup.equals("A-") || donorBloodGroup.equals("B-") || 
                       donorBloodGroup.equals("AB-") || donorBloodGroup.equals("O-");
            case "O+":
                return donorBloodGroup.equals("O+") || donorBloodGroup.equals("O-");
            case "O-":
                return donorBloodGroup.equals("O-");
            default:
                return false;
        }
    }
} 

File: app/src/main/java/com/example/bloodbank/CompatibleUsersActivity.kt
package com.example.bloodbank

import android.os.Bundle
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.bloodbank.Adapter.CompatibleUserAdapter
import com.example.bloodbank.Model.CompatibleUser
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*

/**
 * CompatibleUsersActivity - Display list of compatible blood donors.
 * Simple Firebase query activity, no complex ViewModel needed.
 */
class CompatibleUsersActivity : AppCompatActivity() {
    
    private lateinit var toolbar: Toolbar
    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: CompatibleUserAdapter
    
    private val firebaseAuth = FirebaseAuth.getInstance()
    private val database = FirebaseDatabase.getInstance()
    private val usersList = mutableListOf<CompatibleUser>()
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_compatible_users)
        
        initializeViews()
        setupToolbar()
        setupRecyclerView()
        loadCompatibleUsers()
    }
    
    private fun initializeViews() {
        toolbar = findViewById(R.id.toolbar)
        recyclerView = findViewById(R.id.compatibleUsersRecyclerView)
    }
    
    private fun setupToolbar() {
        setSupportActionBar(toolbar)
        supportActionBar?.apply {
            title = "Compatible Donors"
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowHomeEnabled(true)
        }
    }
    
    private fun setupRecyclerView() {
        adapter = CompatibleUserAdapter(usersList)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = adapter
    }
    
    private fun loadCompatibleUsers() {
        val currentUserId = firebaseAuth.currentUser?.uid ?: return
        
        // Get current user's blood group
        database.reference.child("users").child(currentUserId)
            .addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    val bloodGroup = snapshot.child("bloodgroup").getValue(String::class.java)
                    bloodGroup?.let { loadDonorsByBloodGroup(it) }
                }
                
                override fun onCancelled(error: DatabaseError) {
                    Toast.makeText(
                        this@CompatibleUsersActivity,
                        "Error loading user data",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            })
    }
    
    private fun loadDonorsByBloodGroup(recipientBloodGroup: String) {
        val compatibleBloodGroups = getCompatibleBloodGroups(recipientBloodGroup)
        
        database.reference.child("users")
            .orderByChild("type")
            .equalTo("donor")
            .addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    usersList.clear()
                    
                    snapshot.children.forEach { userSnapshot ->
                        val bloodGroup = userSnapshot.child("bloodgroup").getValue(String::class.java)
                        
                        if (bloodGroup in compatibleBloodGroups) {
                            val user = CompatibleUser(
                                id = userSnapshot.child("id").getValue(String::class.java) ?: "",
                                name = userSnapshot.child("name").getValue(String::class.java) ?: "",
                                bloodgroup = bloodGroup ?: "",
                                phonenumber = userSnapshot.child("phonenumber").getValue(String::class.java) ?: "",
                                address = userSnapshot.child("address").getValue(String::class.java) ?: ""
                            )
                            usersList.add(user)
                        }
                    }
                    
                    adapter.notifyDataSetChanged()
                }
                
                override fun onCancelled(error: DatabaseError) {
                    Toast.makeText(
                        this@CompatibleUsersActivity,
                        "Error loading donors",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            })
    }
    
    private fun getCompatibleBloodGroups(recipientBloodGroup: String): List<String> {
        return when (recipientBloodGroup) {
            "A+" -> listOf("A+", "A-", "O+", "O-")
            "A-" -> listOf("A-", "O-")
            "B+" -> listOf("B+", "B-", "O+", "O-")
            "B-" -> listOf("B-", "O-")
            "AB+" -> listOf("A+", "A-", "B+", "B-", "AB+", "AB-", "O+", "O-")
            "AB-" -> listOf("A-", "B-", "AB-", "O-")
            "O+" -> listOf("O+", "O-")
            "O-" -> listOf("O-")
            else -> emptyList()
        }
    }
    
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                finish()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
}
